doc/type_adaption.md

```diff
@@ -0,0 +1,276 @@
+any type more than once with differing parameters, so we can just choose an
+arbitrary path to follow. This implementation always follows the first child.
+
+It is important to note that this will *not* cause us to get stuck by taking a
+wrong turn and no longer having a path down to our target class. We build the
+hierarchy from the bottom up and therefore ensure that we can get down to where
+we need to go from any other node in the tree.
+
+#### Finding no hierarchy
+If we can not find any hierarchy we currently return the input type unchanged.
+
+
+## Adapting a method to a subclass
+Closely related to type adaption (but not exactly the same!) is translating
+*entire methods*. As a first approach we might try to just re-use our type
+adaption code and adapt the return, parameter and thrown types down to the
+target class. This, however, is not sufficient *as methods might declare their
+own type parameters*. Consider this example:
+```java
+interface Parent<T> {
+  <R extends T> R foo();
+}
+interface Child<Q> extends Parent<Q> {}
+```
+If we translated all types using the type adaption algorithm above we might get
+```java
+R foo();
+```
+as there is no equivalent to `R` we can adapt to - there just is *nothing* here
+which would correspond to the method type parameter with an updated bound.
+
+What we do instead is cloning the method and copying the formal type parameters
+but *adapting their bound*. The example above is first turned into:
+```java
+<R extends Q> R foo();
+```
+After this step the other types are translated using the normal algorithm. As
+we keep the names of the formal type parameters the same we do not need to
+rewrite the method body, but we might incur clashes with existing type variables
+in the subclass (have a look at what happens when you rename `Q` to `T` in the
+subclass and add a `T t` parameter to the superclass method).
+
+
+## Checking for conflicting methods
+It is sometimes necessary to find out whether two methods *conflict*, i.e.
+whether you can not declare *both* in the same type. A simple example would be
+```java
+String foo();
+String foo();
+```
+Obviously you can only have one of those in a type at a given time!
+A simple way to detect this is just checking whether the names and parameter
+types of the methods match.
+
+Sadly, generic erasure makes this problem a lot more difficult. The following
+two methods also conflict:
+```java
+void foo(List<String> list);
+void foo(List<Integer> list);
+```
+One solution to fix this is only looking at the *type erasure* of the two
+methods. And this indeed solves the issue here as the methods would erase to:
+```java
+void foo(List list);
+void foo(List list);
+```
+Those methods have the same name and parameters and therefore obviously conflict.
+
+Alas, this is still not enough. Consider this:
+```java
+interface Foo<T> {
+  void foo(T t);
+}
+interface SubFoo extends Foo<String> {
+  void foo(String t);
+}
+```
```

Comment:
I'm a bit confused here. This section is talking about methods that:

> *conflict*, i.e. whether you can not declare *both* in the same type.

But the above example is fine, it's just an override. What precisely do you mean by "conflict"?

Reply:
This definition:
> It is sometimes necessary to find out whether two methods *conflict*, i.e. whether you can not declare *both* in the same type.

The Javadoc for the method is probably a bit more helpful:
```java
	 * Checks if two given methods are conflicting, i.e. they can not both be declared in the same
	 * class. This happens if the erasure of the methods is the same or one overrides the other. This
	 * method is used to remove methods that were already visited or were overwritten/shadowed by a
	 * subclass in various places.
```
It's currently used in the metamodel classes to figure out if something was inherited (which seems like a weird way to do that, but I didn't touch the Metamodel).  
It is also used in `getAllMethods` to reject duplicated methods (which presumably can occur if a class is visited twice or a superclass method is found after an overriding one in the subclass).